### PR TITLE
Simplified url routing

### PR DESCRIPTION
**Pull request by knbk**
Commits implementing the changes I suggested [here](https://github.com/django/django/pull/7482#pullrequestreview-54542985).

### CODE REVIEW

File: django/urls/resolvers.py [hunk 1]
```diff
@@ -160,41 +163,48 @@ def _route_to_regex(route):
...
-class PathURLMixin:
+class RoutePattern:
+    def __init__(self, route):
+        self.regex = route
+
     def compile(self):
         # TODO: Wrap in `try/except:raise ImproperlyConfigured`
-        regex, converters = _route_to_regex(str(self.pattern))
+        regex, converters = _route_to_regex(str(self.regex))
         if isinstance(self, URLPattern):
```
Comment (sjoerdjob) [inline django/urls/resolvers.py]: This was originally already hacky, but now also broken.

The purpose of this if was to check if it was a final pattern (view) or not (an include). If it was final, we don't want to allow trailing text.

File: django/urls/resolvers.py [hunk 2]
```diff
@@ -160,41 +163,48 @@ def _route_to_regex(route):
     return ''.join(parts), converters
 
 
-class PathURLMixin:
+class RoutePattern:
+    def __init__(self, route):
+        self.regex = route
```
Comment (sjoerdjob) [inline django/urls/resolvers.py]: It's not a regular expression. A "pattern" maybe.

File: django/views/templates/technical_404.html [hunk 3]
```diff
@@ -52,7 +52,7 @@
         {% for pattern in urlpatterns %}
           <li>
             {% for pat in pattern %}
-                {{ pat.pattern }}
+                {{ pat.pattern.regex }}
```
Comment (sjoerdjob) [inline django/views/templates/technical_404.html]: Maybe the `...Pattern` could implement a `__str__` that shows the proper argument? Or an explicit method for "string in technical 404"?

**Review by sjoerdjob — COMMENTED** [review]
It's even more code-churn and removed classes as before. Maybe we could leave RegexURLResolver and RegexURLPattern as passthrough stubs? If that backwards compatibility I'd required. Maybe it isn't.

**Comment by sjoerdjob** [comment]
(and I forgot to say: thanks for the changes.)

**Comment by knbk** [comment]
`RegexURLResolver` has never been a documented API, so backwards compatibility isn't required. Personally I don't think its use in third-party apps is widespread enough to warrant a stub.

**Comment by sjoerdjob** [comment]
I've been thinking about the approach you're taking here, having a bit of mixed feelings about it. Overall, I think it's an improvement.

However, judging the code 'as is' (without looking at what it was or will become) (and take this with a sufficient dosage [a grain] of salt):

I think the `RegexPattern` and `RoutePattern` classes don't really do enough to justify being separate classes (on the other hand: not sure if the mixins were justifiable either, and that also suffered from a potential combinatoric explosion).

It seems to me that there should be more logic moved to these `*Pattern` classes:

1. Matching and conversion could be moved there. (To replace the `match.groupdict()` that's now used in `.resolve()`)
2. It could "check" itself a bit more: for `RegexPattern` the `include()`/`$`-check could now also be implemented using the same `_is_endpoint` (if added to `re_path`/`RegexPattern`). The "starts with slash" check can also be moved to the pattern itself.
3. The approach to handling of translatable patterns was already quite troublesome. I'm not sure that it has not gotten worse by choosing composition over mixins.

Also, based on the `$`-issue: it might be a good idea to add a test that `path('hello/', <callable view>)` does not match `/hello/world/` but only `/hello/`.

Anyway, like I said: it's already an improvement. I'll leave it up to you whether or not to address the points above. If you want I can just merge the commits so that they will also be processed by the CI.

**Comment by sjoerdjob** [comment]
(because of inactivity, I decided to merge what was already there)

**Pull request merged**